pandas/core/frame.py

```diff
@@ -3892,6 +3893,22 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
         if not isinstance(keys, list):
             keys = [keys]
 
+        missing = []
+        for x in keys:
+            if not (is_scalar(x) or isinstance(x, tuple)):
+                if not isinstance(x, (ABCSeries, ABCIndexClass, ABCMultiIndex,
+                                      list, np.ndarray)):
+                    raise TypeError('keys may only contain a combination of '
```

Comment:
this is not true. either all have to be scalars, or all list-likes (yes they don't have to be the same, just list-convertible). 

Please make this more readable from a user perspective. If I got this I would have no idea what to do.

Reply:
See longer comment in the main thread.

Tuples are allowed because they might be valid column labels.